### DISCUSSION TITLE
Argument "bind" to "sessionmaker" has incompatible type ...

### DISCUSSION THREAD

**Discussion by ddrouin** [sqlalchemy/sqlalchemy#9713] [Usage Questions] [answered]
Getting the following typing error from mypy:

`error: Argument "bind" to "sessionmaker" has incompatible type "Union[Engine, AsyncEngine]"; expected "Union[Connection, Engine, None]"  [arg-type] `

Code:

```
        if self.async_mode:
            self.session_class = sessionmaker(bind=self.engine, class_=AsyncSession, expire_on_commit=False)
        else:
            self.session_class = sessionmaker(bind=self.engine)
```

Issue is raised on the else path.  So `self.engine` in our code may wind up as an Engine or AsyncEngine instance depending on whether we'll be running in async mode hence mypy finding that its type is `"Union[Engine, AsyncEngine]"`

Following along with the example: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-orm
```
    engine = create_async_engine(
        "postgresql+asyncpg://scott:tiger@localhost/test",
        echo=True,
    )
    ...
    async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
```

Looks like you should be able to pass in an AsynsEngine instance there.  In looking at the type annotations for sessionmaker:

`class sessionmaker(_SessionClassMethods, Generic[_S]):`

With _S defined as

```
    class_: Type[_S]

    @overload
    def __init__(
        self,
        bind: Optional[_SessionBind] = ...,
        *,
        class_: Type[_S],
        autoflush: bool = ...,
        expire_on_commit: bool = ...,
        info: Optional[_InfoType] = ...,
        **kw: Any,
    ):
        ...

    @overload
    def __init__(
        self: "sessionmaker[Session]",
        bind: Optional[_SessionBind] = ...,
        *,
        autoflush: bool = ...,
        expire_on_commit: bool = ...,
        info: Optional[_InfoType] = ...,
        **kw: Any,
    ):
        ...

    def __init__(
        self,
        bind: Optional[_SessionBind] = None,
        *,
        class_: Type[_S] = Session,  # type: ignore
        autoflush: bool = True,
        expire_on_commit: bool = True,
        info: Optional[_InfoType] = None,
        **kw: Any,
    ):
```

and _SessionBind is defined as:

`_SessionBind = Union["Engine", "Connection"]`

So that explains this part of the mypy error message: `expected "Union[Connection, Engine, None]"`
I'm not seeing then how an AsyncEngine would be acceptable to pass in. Or perhaps I'm doing something incorrectly that's resulting in this error?

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's expected. You should use https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.async_sessionmaker for asyncio

**Comment by ddrouin** [comment]
Note:
https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-orm

Has this:
`from sqlalchemy.orm import sessionmaker`

Which differs from 2.0
https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#synopsis-orm

I'll follow the 2.0 version even though I'm still on 1.4.

**Reply by ddrouin** [reply]
OK now I get: 
`error: Module "sqlalchemy.ext.asyncio" has no attribute "async_sessionmaker"; maybe "async_session"?`
So apparently now something available in 1.4.

**Reply by CaselIT** [reply]
Sorry I assumed 2.0

In 1.4 typing is not great, I would stick a type_ignore there

**Reply by ddrouin** [reply]
yep good call - thanks!

**Reply by ddrouin** [reply]
Just for anyone else that hits this - workaround:

```
       if self.async_mode:
            self.session_class = cast(AsyncSession, sessionmaker(bind=cast(AsyncEngine, self.engine),
                                                                 class_=AsyncSession, expire_on_commit=False))
        else:
            self.session_class = cast(Session, sessionmaker(bind=cast(Engine, self.engine)))
```
